docsk8s/setting-up/ls-k8s-persistent-storage.asciidoc

```diff
@@ -1,16 +1,267 @@
+        labels:
+          app: logstash-demo
+      spec:
+        accessModes: ["ReadWriteOnce"]
+        resources:
+          requests:
+            storage: 2Gi
+--
+
+Everything is similar to `Deployment`, except the usage of `VolumeClaimTemplates`.
+
+<1> Request 2G of persistent storage from `PersistentVolumes`.
+
+<2> Mount the storage to `/usr/share/logstash/data`. This is the default path of {ls} and its plugins for any persistence needs.
+
+NOTE: The feature of persistent volume expansion depends on the storage class, checkout your cloud provider.
+
+[[persistent-storage-pq]]
+==== Persistent queue (PQ)
+
+To enable {logstash-ref}/persistent-queues.html[PQ] for every pipeline, specify options in `logstash.yml`. 
+
+[source,yaml]
+--
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: logstash-config
+data:
+  logstash.yml: |
+    api.http.host: "0.0.0.0"
+    queue.type: persisted
+    queue.max_bytes: 1024mb
+...
+--
+
+To specify options per pipeline, set in `pipelines.yml`.
+
+[source,yaml]
+--
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: logstash-config
+data:
+  logstash.yml: |
+    api.http.host: "0.0.0.0"
+  pipelines.yml: |
+    - pipeline.id: fast_ingestion
+      path.config: "/usr/share/logstash/pipeline/fast.conf"
+      queue.type: persisted
+      queue.max_bytes: 1024mb
+    - pipeline.id: slow_ingestion
+      path.config: "/usr/share/logstash/pipeline/slow.conf"
+      queue.type: persisted
+      queue.max_bytes: 2048mb
+--
+
+Queue data store in `/usr/share/logstash/data/queue` by default.
+
+[[persistent-storage-pq-util]]
+===== Run PQ utility
+
+In case the pod is terminated ungracefully and hence the persistent queues are corrupted, `pqcheck` and `pqrepair` are the tools for troubleshooting.
+
+To run {logstash-ref}/persistent-queues.html#pqcheck[pqcheck] to identify corrupted files: 
+
+[source,bash]
+--
+kubectl exec logstash-0 -it -- /usr/share/logstash/bin/pqcheck /usr/share/logstash/data/queue/pipeline_id
+--
+
+To run {logstash-ref}/persistent-queues.html#pqrepair[pqrepair] to repair the queue: 
+
+[source,bash]
+--
+kubectl exec logstash-0 -it -- /usr/share/logstash/bin/pqrepair /usr/share/logstash/data/queue/pipeline_id
+--
+
+[[persistent-storage-pq-drain]]
+===== Draining the queue
+
+Set `queue.drain: true` to drain and wait for the queues to empty when receive SIGTERM. However, the default `TerminationGracePeriodSeconds` in Kubernetes is 30 seconds. If {ls} is still running after the grace period elapsed, Kubernetes send SIGKILL to terminated the pod immediately. Hence, {ls} cannot finish the draining.
+
+A workaround is to set a very long period like 1 year to `TerminationGracePeriodSeconds` to make sure {ls} get enough time to drain all events.
+
+[[persistent-storage-dlq]]
+==== Dead letter queue (DLQ)
+
+To enable {logstash-ref}/dead-letter-queues.html[dead letter queue], specify options in `logstash.yml`. The default path of DLQ is `/usr/share/logstash/data/dead_letter_queue`.
```

Comment:
Do we need to mention that the queue contents themselves will be in `/usr/share/logstash/data/dead_letter_queue/<pipeline_id>`

Reply:
probably no. When we config a different path for DLQ, we also did not include <pipeline_id>.